compiler/rustc_middle/src/traits/mod.rs

```diff
@@ -651,41 +651,29 @@ pub enum ImplSource<'tcx, N> {
     /// any).
     Param(Vec<N>, ty::BoundConstness),
 
-    /// Virtual calls through an object.
-    Object(ImplSourceObjectData<N>),
-
-    /// Successful resolution for a builtin trait.
-    Builtin(Vec<N>),
-
-    /// ImplSource for trait upcasting coercion
-    TraitUpcasting(ImplSourceTraitUpcastingData<N>),
+    /// Successful resolution for a builtin impl.
+    Builtin(BuiltinImplSource, Vec<N>),
 }
 
 impl<'tcx, N> ImplSource<'tcx, N> {
     pub fn nested_obligations(self) -> Vec<N> {
         match self {
             ImplSource::UserDefined(i) => i.nested,
-            ImplSource::Param(n, _) | ImplSource::Builtin(n) => n,
-            ImplSource::Object(d) => d.nested,
-            ImplSource::TraitUpcasting(d) => d.nested,
+            ImplSource::Param(n, _) | ImplSource::Builtin(_, n) => n,
```

Comment:
nit: maybe `Param` and `Builtin` should agree with their order of `nested` and the remaining data, I would change the order of `Param` xd